x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidator.java

```diff
@@ -0,0 +1,513 @@
+        boolean isDeferrable();
+
+        void validate(Context context);
+    }
+
+    // note: this is equivalent to the default for search requests
+    private static final IndicesOptions DEFAULT_INDICES_OPTIONS_FOR_VALIDATION = IndicesOptions
+        .strictExpandOpenAndForbidClosedIgnoreThrottled();
+
+    private static final SourceDestValidation SOURCE_MISSING_VALIDATION = new SourceMissingValidation();
+    private static final SourceDestValidation REMOTE_SOURCE_VALIDATION = new RemoteSourceEnabledAndRemoteLicenseValidation();
+
+    private static final List<SourceDestValidation> PREVIEW_VALIDATIONS = Arrays.asList(
+        SOURCE_MISSING_VALIDATION,
+        REMOTE_SOURCE_VALIDATION
+    );
+
+    private static final List<SourceDestValidation> START_AND_PUT_VALIDATIONS = Arrays.asList(
+        SOURCE_MISSING_VALIDATION,
+        REMOTE_SOURCE_VALIDATION,
+        new DestinationInSourceValidation(),
+        new DestinationSingleIndexValidation()
+    );
+
+    /**
+     * Validate dest request.
+     *
+     * This runs a couple of simple validations at request time, to be executed from a {@link ActionRequest}}
+     * implementation.
+     *
+     * Note: Source can not be validated at request time as it might contain expressions.
+     *
+     * @param validationException an ActionRequestValidationException for collection validation problem, can be null
+     * @param dest destination index, null if validation shall be skipped
+     */
+    public static ActionRequestValidationException validateRequest(
+        @Nullable ActionRequestValidationException validationException,
+        @Nullable String dest
+    ) {
+        try {
+            if (dest != null) {
+                validateIndexOrAliasName(dest, InvalidIndexNameException::new);
+                if (dest.toLowerCase(Locale.ROOT).equals(dest) == false) {
+                    validationException = addValidationError(getMessage(DEST_LOWERCASE, dest), validationException);
+                }
+            }
+        } catch (InvalidIndexNameException ex) {
+            validationException = addValidationError(ex.getMessage(), validationException);
+        }
+
+        return validationException;
+    }
+
+    /**
+     * Validates source and dest indices for preview.
+     *
+     * @param clusterState The current ClusterState
+     * @param indexNameExpressionResolver A valid IndexNameExpressionResolver object
+     * @param remoteClusterService A valid RemoteClusterService object
+     * @param remoteClusterLicenseChecker A RemoteClusterLicenseChecker or null if CCS is disabled
+     * @param source an array of source indices
+     * @param dest destination index
+     * @param nodeName the name of this node
+     * @param license the license of the feature validated for
+     * @return ValidationException instance containing all validation errors or null if validation found no problems.
+     */
+    public static ValidationException validateForPreview(
```

Comment:
As discussed, this seems like we should not have this and have the method caller pass in their desired validations.